File: frontend/src/mateu/shared/apiClients/dtos/Card.ts
import ComponentMetadata from "./ComponentMetadata";
import {CardLayout} from "./CardLayout";
import Action from "./Action";
import {ServerSideObject} from "./ServerSideObject";

export default interface Card extends ComponentMetadata {

    layout: CardLayout
    thumbnail: string
    headerText: string
    subhead: string
    media: string
    supportingText: string
    buttons: Action[]
    icons: Action[]
    actionHandler: ServerSideObject

}

File: backend/core/src/main/java/io/mateu/core/domain/uidefinition/core/interfaces/Card.java
package io.mateu.core.domain.uidefinition.core.interfaces;

import io.mateu.core.domain.uidefinition.shared.interfaces.ActionHandler;
import io.mateu.dtos.ServerSideObject;

import java.util.Collections;
import java.util.List;

public record Card(
    CardLayout layout,
    String thumbnail,
    String headerText,
    String subhead,
    String media,
    String supportingText,
    List<Button> buttons,
    List<CallableIcon> icons,
    ActionHandler actionHandler) {

  public Card(String headerText, String subhead, String supportingText) {
    this(CardLayout.Layout1, "", headerText, subhead, "", supportingText, List.of(), List.of(), null);
  }

  public Card {
    buttons = buttons != null ? Collections.unmodifiableList(buttons) : List.of();
    icons = icons != null ? Collections.unmodifiableList(icons) : List.of();
  }

  @Override
  public List<Button> buttons() {
    return Collections.unmodifiableList(buttons);
  }

  @Override
  public List<CallableIcon> icons() {
    return Collections.unmodifiableList(icons);
  }
}
